File: stack/imp_arr.py
class Stack:
    def __init__(self, size):
        self.size = size
        self.arr = [None] * size
        self.top = -1

    def isEmpty(self):
        return self.top == -1

    def isFull(self):
        return self.top == self.size - 1

    def push(self, x):
        if self.isFull():
            print("Stack Overflow")
        else:
            self.top += 1
            self.arr[self.top] = x

    def pop(self):
        if self.isEmpty():
            print("Stack Underflow")
        else:
            val = self.arr[self.top]
            self.top -= 1
            return val

    def peek(self):
        if self.isEmpty():
            print("Stack is Empty")
        else:
            return self.arr[self.top]

    def display(self):
        if self.isEmpty():
            print("Stack is Empty")
        else:
            for i in range(self.top, -1, -1):
                print(self.arr[i], end=" ")
            print()
s = Stack(5)
s.push(10)
s.push(20)
s.push(30)
s.display()      
print(s.pop())   
print(s.peek())  
s.display()      

File: stack/imp_arr.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 100

typedef struct S {
    int a[MAX]; 
    int t;     
} S;

void init(S* s) {
    s->t = -1;
}

void push(S* s, int x) {
    if (s->t == MAX - 1) {
        printf("Stack Overflow\n");
        return;
    }
    s->a[++s->t] = x;
}

int pop(S* s) {
    if (s->t == -1) {
        printf("Stack Underflow\n");
        return -1;
    }
    return s->a[s->t--];
}

int peek(S* s) {
    return s->t == -1 ? -1 : s->a[s->t];
}

int empty(S* s) {
    return s->t == -1;
}


void disp(S* s) {
    for (int i = s->t; i >= 0; i--)
        printf("%d ", s->a[i]);
    printf("\n");
}

int main() {
    S stk;
    init(&stk);

    push(&stk, 10);
    push(&stk, 20);
    push(&stk, 30);

    disp(&stk);        
    printf("%d\n", pop(&stk));  
    printf("%d\n", peek(&stk)); 
    disp(&stk);       

    return 0;
}
